lib/puppet/provider/network_trunk/cisco_nexus.rb

```diff
@@ -0,0 +1,127 @@
+require 'puppet/resource_api/simple_provider'
+
+# Implementation for the network_trunk type using the Resource API.
+class Puppet::Provider::NetworkTrunk::CiscoNexus < Puppet::ResourceApi::SimpleProvider
+  def canonicalize(_context, resources)
+    resources.each do |resource|
+      resource[:tagged_vlans] = resource[:tagged_vlans].sort_by(&:to_i) if resource[:tagged_vlans]
+    end
+    resources
+  end
+
+  def get(_context, interface_names=nil)
+    current_states = []
+    if interface_names.nil?
+      @interfaces ||= Cisco::Interface.interfaces
+      @interfaces.each do |interface_name, i|
+        next unless i.send(:switchport_mode) == :trunk
+        array_vlans = convert_allowed_vlan_to_array(i.send(:switchport_trunk_allowed_vlan))
+        interface = {
+          name:          interface_name,
+          untagged_vlan: i.send(:switchport_trunk_native_vlan),
+          tagged_vlans:  array_vlans.sort_by(&:to_i),
+          mode:          'trunk',
+          ensure:        'present',
+        }
+        current_states << interface
+      end
+    else
+      interface_names.each do |interface_name|
+        @interfaces ||= Cisco::Interface.interfaces
+        interface = @interfaces[interface_name]
+        return [] unless interface.send(:switchport_mode) == :trunk
+        array_vlans = convert_allowed_vlan_to_array(interface.send(:switchport_trunk_allowed_vlan))
+
+        interface_result = {
+          name:          interface_name,
+          untagged_vlan: interface.send(:switchport_trunk_native_vlan),
+          tagged_vlans:  array_vlans.sort_by(&:to_i),
+          mode:          'trunk',
+          ensure:        'present',
+        }
+        current_states << interface_result
+      end
+    end
+    current_states
+  end
+
+  def update(context, name, should)
+    validate_should(should)
+    context.notice("Updating '#{name}' with #{should.inspect}")
+    @interfaces ||= Cisco::Interface.interfaces
+    @interfaces[name].switchport_mode = should[:mode].to_sym if should[:mode]
+    @interfaces[name].switchport_trunk_native_vlan = should[:untagged_vlan] if should[:untagged_vlan]
+    @interfaces[name].switchport_trunk_allowed_vlan = convert_array_to_allowed_vlan(should[:tagged_vlans]) if should[:tagged_vlans]
+  end
+
+  def create(context, name, should)
+    validate_should(should)
+    context.notice("Creating '#{name}' with #{should.inspect}")
+    Cisco::Interface.new(name)
+    @interfaces ||= Cisco::Interface.interfaces
+    @interfaces[name].switchport_mode = should[:mode].to_sym if should[:mode]
+    @interfaces[name].switchport_trunk_native_vlan = should[:untagged_vlan] if should[:untagged_vlan]
+    @interfaces[name].switchport_trunk_allowed_vlan = convert_array_to_allowed_vlan(should[:tagged_vlans]) if should[:tagged_vlans]
+  end
```

Comment:
Yep, we do that in a few places in Cisco_IOS for same create/update.

Your only difference is the context logging, either make this generic or axe it?

Reply:
Another option is to skip the simple provider and just call a single function (update?) + delete, and then keep the logging within the set. You don't use the context otherwise, wouldn't even need to be passed in in this case.

Reply:
Well updating suits even if "creating" as technically the interface is never destroyed, just the "mode" realistically, so if ethernet1/1 is currently trunk, then we destroy it, ethernet1/1 still exists, so if we "create" it again, we're just updating its attributes. So I think the `alias create update` would be fine with the context remaining as `updating`

Reply:
https://github.com/puppetlabs/cisco_ios/blob/3c0ea3dd19b123f5d94130c31fed0f9c31d92399/lib/puppet/provider/network_trunk/ios.rb#L87

for ref

Reply:
Alternatively an example of a custom set w/o simpleprovider

https://github.com/puppetlabs/cisco_ios/blob/3c0ea3dd19b123f5d94130c31fed0f9c31d92399/lib/puppet/provider/network_interface/ios.rb#L121